### ISSUE TITLE
DelegatingCompletableFuture does not complete if the callback throws an exception

### ISSUE THREAD

**Issue by dsyer** [hazelcast/hazelcast#18053]
From the javadocs for `CompletionStage.whenComplete()`:

```
Unlike method handle, this method is not designed to translate completion outcomes, so the supplied action 
should not throw an exception. However, if it does, the following rules apply: if this stage completed normally 
but the supplied action throws an exception, then the returned stage completes exceptionally with the supplied 
action's exception. Or, if this stage completed exceptionally and the supplied action throws an exception, then 
the returned stage completes exceptionally with this stage's exception.
```

`DelegatingCompletableFuture` does not implement this contract. In particular the `DeserializingBiConsumer` is prone to throwing `HazelcastSerializationException` which then causes downstream consumers of the future to hang because there is no completion signal.

**Comment by vbekiaris** [comment]
@dsyer thanks for the issue report. I tried to reproduce the issue with the chained future not being completed when `HazelcastSerializationException` is thrown but failed to do so. My test looks like this:

```
    // SerializationService#toObject will throw HazelcastSerializationException
    private final Data invalid = new HeapData(new byte[] {0, 0, 0, 0, 5, 0, 0, 0, 0});
    ...
    @Test
    public void testWhenComplete() {
        CompletableFuture<Object> future = new DelegatingCompletableFuture<>(serializationService,
                                       InternalCompletableFuture.newCompletedFuture(invalid));
        // lambda will be wrapped in `DeserializingBiConsumer` and will attempt to deserialize invalid data
        CompletableFuture<Object> chained = future.whenComplete((v, t) -> ignore());

        Throwable t = assertThrows(CompletionException.class, chained::join);
        assertInstanceOf(HazelcastSerializationException.class, t.getCause());
    }
```

The `HazelcastSerializationException` is actually caught by the delegate future's `whenComplete` implementation (eg when `DelegatingCompletableFuture` wraps a Hazelcast invocation, this happens in `AbstractInvocationFuture#unblockWhenComplete` or `AbstractInvocationFuture$WhenCompleteNode`).

Can you share a specific use case that fails for you?

**Comment by dsyer** [comment]
The assertion contradicts the contract, doesn't it, because it ignores the failure? The problem is that the user's chained callback never gets the signal.

A better test would be (without actually compiling and running it):

```java
    // SerializationService#toObject will throw HazelcastSerializationException
    private final Data invalid = new HeapData(new byte[] {0, 0, 0, 0, 5, 0, 0, 0, 0});
    ...
    @Test
    public void testWhenComplete() {
        CompletableFuture<Object> future = new DelegatingCompletableFuture<>(serializationService,
                                       InternalCompletableFuture.newCompletedFuture(invalid));
        // lambda will be wrapped in `DeserializingBiConsumer` and will attempt to deserialize invalid data
        AtomicReference<Throwable> error = new AtomicReference<>();
        CompletableFuture<Object> chained = future.whenComplete((v, t) -> error.set(t));

        assertThrows(CompletionException.class, chained::join);
        // but the exception is still passed on to the user action
        assertInstanceOf(HazelcastSerializationException.class, error.get());
    }
```

**Comment by vbekiaris** [comment]
> The problem is that the user's chained callback never gets the signal.

Thanks for the clarification, I see what you mean. I am working on a fix and will keep you posted.

**Comment by vbekiaris** [comment]
@dsyer thanks for the issue report. The fix is now merged in `4.2-SNAPSHOT` and will be backported to `4.0.z` and `4.1.z`.
